src/lib/grad.jl

```diff
@@ -73,8 +73,7 @@ julia> hessian(sin, pi/2)
 """
 hessian(f, x) = hessian_dual(f, x)
 
-hessian_dual(f, x::AbstractArray) = forward_jacobian(x -> gradient(f, x)[1], x)[2]
-
+hessian_dual(f, x::AbstractArray) = ForwardDiff.jacobian(x -> gradient(f, x)[1], x)
```

Comment:
The example you test relies on the gradient rule for ForwardDiff.jacobian, which assumes its `f` is pure. This is true here.

It would be worth trying to look at some examples where this isn't true, i.e. `f` closes over some parameters (either arrays or numbers). In this case their gradients are not traced. I believe you should get a warning from ForwardDiff.jacobian's rule. 

Are there any further surprises? Anything which could be tested?

Reply:
Did you mean nn layers from Flux? I will dig in a bit tomorrow.

The scalar indexing in the above comment can be added to testing as well.